context_drm_gl: add support for hdr metadata

Dudemanguy opened the pull request:
Somebody please actually test since I have no actual hdr capable monitors. There's also a `HDMI_EOTF_TRADITIONAL_GAMMA_HDR` enum in there but I have no idea what on earth that was supposed to be so I ignored it.

Also not sure if any existing `--drm-format` options are capable enough here.

kasper93 commented on video/out/drm_common.c:
```diff
@@ -1240,6 +1281,43 @@ double vo_drm_get_display_fps(struct vo_drm_state *drm)
+    destroy_hdr_blob(drm);
+    drm->target_params = target_params;
+
+    struct pl_hdr_metadata hdr = target_params.color.hdr;
```
```suggestion
    const struct pl_hdr_metadata *hdr = &target_params.color.hdr;
```
no need to copy.

Also we should check if colorspace is hdr before sending it, I think. 

Also 
> Userspace first need to detect the HDR capabilities of sink by reading and parsing the EDID.

I don't think we do that? Probably nothing terrible would happen when we send blob to unsupported display, but what if?

Dudemanguy replied:
We do need to send it in the sdr case so the display can unset itself if the playlist switches from an hdr file to an sdr one. It is not like we are sending anything wrong so it should be OK.

>Probably nothing terrible would happen when we send blob to unsupported display, but what if?

Well just nothing happens but it's probably worth checking so I'll add that.

Dudemanguy commented:
I added some sanity checking to this by using `libdisplay-info`. Only bothered with static metadata for now. In theory if the display doesn't support what the target_params specify it should not attempt to use it and fallback to sdr mode.

juw commented:
I tested this with a HDR (DisplayHDR 600) capable monitor and the monitor confirmed it went into HDR mode. It seems to work fine. I used the following config:
```
vo=gpu-next
drm-connector=HDMI-A-1
drm-format=xrgb2101010
target-colorspace-hint
target-prim=dci-p3
target-trc=pq
target-peak=600
```

Dudemanguy commented:
Off the top of my head, I still should add a check to make sure the monitor supports bt 2020 to this. I'm not sure if there's anything else that would be good to check.

quietvoid commented:
For TVs to playback bt.2020 correctly, the `Colorspace` prop must be set if the connector supports it in the colorimetry data block (EDID).

Dudemanguy commented:
>For TVs to playback bt.2020 correctly, the Colorspace prop must be set if the connector supports it in the colorimetry data block (EDID).

Made an attempt at this. There's quite a few colorspaces but it seems that mutter and kwin both only care about bt2020_rgb so that is what I copied as well. Basically if we get an hdr one, set the colorspace property to bt2020_rgb. If not, set it back to the default one. Not sure if this is strictly correct of course, but there doesn't seem to be much prior art on this.

juw commented:
Hmm, with these new changes, my monitor does claim to go into HDR mode anymore, although it has `BT2020RGB` in its colorimetry data block:
```
$ edid-decode /sys/class/drm/card0-HDMI-A-1/edid
[...]
Block 1, CTA-861 Extension Block:
[...]
  Colorimetry Data Block:
    xvYCC601
    xvYCC709
    BT2020cYCC
    BT2020YCC
    BT2020RGB
    Gamut Boundary Description Metadata Profile P0
[...]
  HDR Static Metadata Data Block:
    Electro optical transfer functions:
      Traditional gamma - SDR luminance range
      Traditional gamma - HDR luminance range
      SMPTE ST2084
    Supported static metadata descriptors:
      Static metadata type 1
    Desired content max luminance: 118 (644.196 cd/m^2)
    Desired content max frame-average luminance: 96 (400.000 cd/m^2)
    Desired content min luminance: 23 (0.052 cd/m^2)
```

Dudemanguy commented:
Does it work again if you comment out these two lines?

```
diff --git a/video/out/drm_common.c b/video/out/drm_common.c
index 938bdfe84c..25e8cd9661 100644
--- a/video/out/drm_common.c
+++ b/video/out/drm_common.c
@@ -694,8 +694,8 @@ static bool target_params_supported_by_display(struct vo_drm_state *drm)
     if (!pl_color_transfer_is_hdr(trc) && !hdr_static_metadata->eotfs->traditional_sdr)
         return false;
 
-    if (pl_color_transfer_is_hdr(trc) && !drm->colorimetry->bt2020_rgb)
-        return false;
+//    if (pl_color_transfer_is_hdr(trc) && !drm->colorimetry->bt2020_rgb)
+//        return false;
 
     if (trc == PL_COLOR_TRC_PQ && !hdr_static_metadata->eotfs->pq)
         return false;
```

juw commented:
I had the same thought, unfortunately no. The only new message in the debug log I get is
```
[vo/gpu-next] Failed to commit atomic request: Error number 22 occurred
```
which points to
```
drm_object_set_property(atomic_ctx->request, atomic_ctx->connector, "COLORSPACE", colorspace);
```
and indead commenting out setting the colorspace makes it work again
```
--- a/video/out/drm_common.c
+++ b/video/out/drm_common.c
@@ -1384,7 +1384,7 @@ bool vo_drm_set_hdr_metadata(struct vo *vo)
         return false;

     int colorspace = pl_color_space_is_hdr(&drm->target_params.color) ? DRM_COLORSPACE_BT2020_RGB : DRM_COLORSPACE_DEFAULT;
-    drm_object_set_property(atomic_ctx->request, atomic_ctx->connector, "COLORSPACE", colorspace);
+    //drm_object_set_property(atomic_ctx->request, atomic_ctx->connector, "COLORSPACE", colorspace);

     const struct pl_hdr_metadata *hdr = &target_params.color.hdr;
     struct hdr_output_metadata metadata = {
```

Dudemanguy commented:
Thanks. Does it work again now?

juw commented:
Ok, so the problem seems to be, that in the newest kernel the enum for `DRM_MODE_COLORIMETRY_BT2020_RGB` is different, then what was defined in your patch, it should be `9` (see https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/drm/drm_connector.h?h=v6.12-rc4#n544) but in here, it evaluated to `6`. Using `9` makes things work:
```
--- a/video/out/drm_common.c
+++ b/video/out/drm_common.c
@@ -1384,7 +1384,7 @@ bool vo_drm_set_hdr_metadata(struct vo *vo)
         return false;

     int colorspace = pl_color_space_is_hdr(&drm->target_params.color) ? DRM_COLORSPACE_BT2020_RGB : DRM_COLORSPACE_DEFAULT;
-    drm_object_set_property(atomic_ctx->request, atomic_ctx->connector, "COLORSPACE", DRM_COLORSPACE_BT2020_RGB);
+    drm_object_set_property(atomic_ctx->request, atomic_ctx->connector, "COLORSPACE", 9);

     const struct pl_hdr_metadata *hdr = &target_params.color.hdr;
     struct hdr_output_metadata metadata = {
```

juw commented:
> Thanks. Does it work again now?

No, with these revised patches, it does only print the `Error number 22` a lot of times more. Setting the colorspace in `vo_drm_set_hdr_metadata` works, as soon as the definition of `DRM_COLORSPACE_BT2020_RGB` matches the definition of `DRM_MODE_COLORIMETRY_BT2020_RGB = 9` as in  the `drm_colorspace` `enum`. in `drm_connector.h`. But at least for my machine it does not seem to be neccessary to set this property at all but the driver seems to figure out what to do by itself.

quietvoid commented:
Yeah the current enum is not correct, `DRM_MODE_COLORIMETRY_BT2020_RGB` is 9.
On my side changing to 9 works fine.

Dudemanguy commented:
Ah thanks for that. I naively assumed the order they were in the drm database was the actual order of the enums in the kernel. Hopefully fixed now.

quietvoid commented:
I'm not exactly sure how `gpu-next` behaves wrt. smaller gamuts in bt.2020 when someone specifies `target-prim=display-p3` or a smaller gamut than bt.2020.
But in almost all cases displays expect to receive bt.2020 so that's the only relevant colorspace to set.
Though there are displays that can receive straight P3 too, but would need more complex colorimetry+Colorspace setting.'

Anyways it works fine on my TV, with the colorimetry passing correctly.
Driver sends out YCbCr 4:4:4.

juw commented:
At least for me, I have to set `target-prim=dci-p3` to get the colours right, when not doing it, `bt.709` is used by default. Using `target-prim=bt.2020` gives wrong colours for the display in use. So `gpu-next` seems to do the right thing with respect to the gamut.

This is what was set (see https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c#n7410) for a device connected via HDMI:
```
kernel: HDR SB:22 02 00 d0 84 80 3e c2 33 c4 86 4c 1d b8 0b 54
kernel: HDR SB:3d 8e 44 58 02 00 00 00 00 00 00 00 00 00 00 00
```
This is for one connected via eDP:
```
kernel: HDR SB:01 1a 02 00 d0 84 80 3e c2 33 c4 86 4c 1d b8 0b
kernel: HDR SB:54 3d 8e 44 58 02 00 00 00 00 00 00 00 00 00 00
```

quietvoid commented:
On my TV, using `target-prim=display-p3` is oversaturated. It only displays properly when sending `bt.2020`.
The option I was looking for was `target-gamut` (https://github.com/mpv-player/mpv/issues/12777) but it doesn't seem to make any difference here.
Overriding the colorimetry on the TV also makes no difference.

On my P3 monitor, `bt.2020` is undersaturated so  I have to use `display-p3`.

Dudemanguy commented:
So I suppose a future direction would be figuring out what `target-prim` is right for the connector and then adding someway to communicate that to vo_gpu_next. That can probably be another PR I think.

Could you guys post the output of this patch applied to the PR real quick on your devices? If we're lucky, this will be enough to detect this.
```
diff --git a/video/out/drm_common.c b/video/out/drm_common.c
index 76a93b2bd3..1b37ba23d9 100644
--- a/video/out/drm_common.c
+++ b/video/out/drm_common.c
@@ -678,6 +678,9 @@ static void setup_edid(struct vo_drm_state *drm)
         }
     }
 
+    if (drm->colorimetry)
+        MP_WARN(drm, "Connector supports bt2020_rgb: %d and st2113: %d\n", drm->colorimetry->bt2020_rgb, drm->colorimetry->st2113_rgb);
+
 done:
     if (blob)
         drmModeFreePropertyBlob(blob);
```

quietvoid commented:
Consumer displays supporting anything other than bt.2020 are pretty rare.
Neither of mine support it, just both `BT2020YCC` and `BT2020RGB`.

Dudemanguy commented:
Hmm that's unfortunate, I was hoping something that needed display-p3 would mark that boolean. So I guess the only assumption we could do in this case is TVs need bt.2020 and monitors need display-p3? That seems pretty fragile. Maybe we just have to let the user set it themselves.

kasper93 commented on meson.build:
```diff
@@ -935,9 +935,11 @@ if features['direct3d']
 
 drm = dependency('libdrm', version: '>= 2.4.105', required: get_option('drm'))
-features += {'drm': drm.found() and (features['vt.h'] or features['consio.h'] or features['wsdisplay-usl-io.h'])}
+libdisplay_info = dependency('libdisplay-info', required: get_option('drm'))
```
Would it be useful to parse EDID and use that info also in non-drm scenarios?

llyyr replied:
When would edid data be useful for us if we aren't the ones modesetting?

Dudemanguy replied:
I can't think of one. On the wayland side, it's actually possible to get most of this information via `xx-color-management-v4` anyways.

kasper93 commented on meson.build:
```diff
@@ -935,9 +935,11 @@ if features['direct3d']
 drm = dependency('libdrm', version: '>= 2.4.105', required: get_option('drm'))
-features += {'drm': drm.found() and (features['vt.h'] or features['consio.h'] or features['wsdisplay-usl-io.h'])}
+libdisplay_info = dependency('libdisplay-info', required: get_option('drm'))
+features += {'drm': drm.found() and libdisplay_info.found() and
```
So, it is mandatory dependency for drm now? Is it ok in practice?

Dudemanguy replied:
All distros are forced to carry this dependency for gnome and kde anyways. Should be fine.

llyyr replied:
Would be useful to check what minimum version we can support

Dudemanguy replied:
It's a bit tricky here since the API technically isn't stable yet and may break in the future but I think for simplicity we can just set 0.2.0 for now. This PR doesn't use it but there are several `di_info_get` functions that were introduced and may potentially be useful.

Dudemanguy replied:
Turns out the pkgconfig version is 0.1.1 so I set it to that.

kasper93 commented on video/out/drm_common.c:
```diff
@@ -588,6 +628,84 @@ static bool setup_crtc(struct vo_drm_state *drm, const drmModeRes *res)
 
+static void setup_edid(struct vo_drm_state *drm)
+{
+    drmModePropertyBlobRes *blob;
```
This needs to be initialized for the loop below to work correctly. (you can also reorder loop body to avoid need to init)

Dudemanguy replied:
I made explicitly made it NULL. Not sure how it could be reordered since I need to use the blob later?

kasper93 commented on video/out/drm_common.c:
```diff
@@ -588,6 +628,84 @@ static bool setup_crtc(struct vo_drm_state *drm, const drmModeRes *res)
+    drm->chromaticity = di_edid_get_chromaticity_coords(edid);
+
+    const struct di_edid_cta *cta;
+    const struct di_edid_ext *const *exts = di_edid_get_extensions(edid);
```
Can `exts == null`? if so this needs to be checked also.

kasper93 commented on video/out/drm_common.c:
```diff
@@ -588,6 +628,84 @@ static bool setup_crtc(struct vo_drm_state *drm, const drmModeRes *res)
+
+    const struct di_edid_cta *cta;
+    const struct di_edid_ext *const *exts = di_edid_get_extensions(edid);
+    for (int i = 0; exts[i]; ++i) {
+        if ((cta = di_edid_ext_get_cta(exts[i])))
+            break;
+    }
```
It is fine, but probably something like would work too.
```suggestion
    while (*exts && !(cta = di_edid_ext_get_cta(*exts++)));
```

Dudemanguy replied:
That should work.

kasper93 commented on video/out/drm_common.c:
```diff
@@ -588,6 +628,84 @@ static bool setup_crtc(struct vo_drm_state *drm, const drmModeRes *res)
+            break;
+    }
+
+    if (cta) {
```
```suggestion
    if (!cta) {
        MP_VERBOSE(drm, "...");
        goto done;
    }
```

kasper93 commented on video/out/drm_common.c:
```diff
@@ -1047,6 +1167,12 @@ void vo_drm_uninit(struct vo *vo)
     if (!drm)
         return;
 
+    restore_sdr(drm);
```
Do we need "restore" function? Wouldn't be better to always call `vo_drm_set_metadata` and act accordingly, set hdr if possible, set sdr otherwise?

Dudemanguy replied:
Well yeah we do have to restore back to sdr on quit otherwise the display will be stuck in hdr mode. We can't rely on the next page flip here either so the atomic commit has also be done in the same step. But probably what it should be is that `restore_sdr` calls `vo_drm_set_hdr_metadata` instead and forces sdr.

Dudemanguy replied:
Refactored the `restore_sdr` and `vo_drm_set_hdr_metadata` logic so there's less redundancy now.

kasper93 commented on video/out/drm_common.c:
```diff
@@ -1240,6 +1365,55 @@ double vo_drm_get_display_fps(struct vo_drm_state *drm)
+        restore_sdr(drm);
+        return false;
+    }
+
```
one line is enough

kasper93 commented on video/out/drm_common.c:
```diff
@@ -1240,6 +1365,55 @@ double vo_drm_get_display_fps(struct vo_drm_state *drm)
+{
+    struct vo_drm_state *drm = vo->drm;
+    struct mp_image_params target_params = vo_get_target_params(vo);
+    if (!vo->target_params || pl_color_space_equal(&target_params.color, &drm->target_params.color))
```
`!vo->target_params` this is not correct, you should check if aquired target params (in line above) have data.

kasper93 commented on video/out/drm_common.c:
```diff
@@ -1240,6 +1365,55 @@ double vo_drm_get_display_fps(struct vo_drm_state *drm)
+    }
+
+
+    struct drm_atomic_context *atomic_ctx = drm->atomic_context;
```
Shouldn't this be checked first?

Dudemanguy replied:
It doesn't need to be checked at all actually. We always have an atomic context here.

kasper93 commented on video/out/drm_common.c:
```diff
@@ -1240,6 +1365,55 @@ double vo_drm_get_display_fps(struct vo_drm_state *drm)
+
+        .hdmi_metadata_type1.eotf = eotf_map[target_params.color.transfer],
+
+        .hdmi_metadata_type1.display_primaries[0].x = roundf(hdr->prim.red.x * DRM_PRIM_FACTOR),
```
`lrintf`?

kasper93 commented on video/out/drm_common.h:
```diff
@@ -31,6 +37,41 @@ enum {
     DRM_OPTS_FORMAT_YUYV,
 };
 
+// Copied from include/linux/hdmi.h
```
I would just copy it verbatim, without changing the name.

Dudemanguy replied:
Should I do it for all of these? I kind of avoided it to shorten the names a bit but I guess maybe it doesn't matter.

kasper93 replied:
I would do it for all of them. Easier to search for reference on those enums, no point if renaming.

Dudemanguy replied:
Done now.

kasper93 commented on video/out/drm_common.h:
```diff
@@ -31,6 +33,41 @@ enum {
     DRM_OPTS_FORMAT_YUYV,
 };
 
+// Copied from include/linux/hdmi.h
```
Are you saying this makes vo_drm GPL now?

Dudemanguy replied:
I am no lawyer but I really doubt some enums mean anything. gstreamer uses several of these and it is also lgpl.

philipl replied:
Indeed. The FSF has explicitly argued that interoperability is fair use, and necessary for various GNU things to work. Enum values won't trigger GPL requirements.

Dudemanguy replied:
Well now this makes even less sense to me. Why would we include a copyright notice solely because of some enums? That's just confusing.

llyyr replied:
These enums are only used for interoperability, this shouldn't infect a project to be under GPL, similar to how linking to a GPL library doesn't make your project GPL. Can you cite some example? This betrays my understanding of licenses altogether. I also see no point

Dudemanguy replied:
What requirements are you even talking about? The hdmi.h one says:

>The above copyright notice and this permission notice (including the
next paragraph) shall be included in all copies or substantial portions
of the Software.

This isn't a copy and this is absolutely not "substantial portions" of the software either.

Dudemanguy replied:
OK well I reconsider when the linux foundation decides to sue gstreamer, gamescope, etc. This is just absurd.

kasper93 commented on video/out/opengl/context_drm_egl.c:
```diff
@@ -319,12 +319,15 @@ static void queue_flip(struct ra_ctx *ctx, struct gbm_frame *frame)
 
-    int ret = drmModeAtomicCommit(drm->fd, atomic_ctx->request,
-                                  DRM_MODE_ATOMIC_NONBLOCK | DRM_MODE_PAGE_FLIP_EVENT, drm);
+    if (vo_drm_set_hdr_metadata(ctx->vo, false))
```
Not really. Not sending color information will produce incorrect output on display.

Dudemanguy replied:
It should unset itself if mpv is terminated properly. Of course we can't do anything about a SIGKILL.

I can make this respect `--target-colorspace-hint` but probably we should just set that option to true by default I think.

kasper93 replied:
There is misconception about what `--target-colorspace-hint` does. It is not meant to control signaling to the display, it is meant to automatically set `--target-` if possible, signaling colorspace should be done unconditionally. Note that it controls how video is rendered to ensure it always is correct. Skipping signaling part in drm layer is too late to do. 

Also I already said multiple times there has to be negotiation between output layer and gpu-next.c to negotiate correct target-colorspace-hint behavior. It may resolve your concern. Currently this PR is impossible to use, because you have to manually set each and all HDR parameters. And even so, it is currently not possible to pass HDR metadata. Simply because how gpu-next is wired up, `--target-trc=pq` will use 0/10000 HDR10 metadata. Only way is to support this property.

So in fact I agree `--target-colorspace-hint` has to be supported/fixed, not in drm.c, but in gpu-next.c. As shortly described in https://github.com/mpv-player/mpv/pull/15048#discussion_r1798704514

> The current solution only works with a limited set of color gamut and features

So it won't do anything for others, nothing breaks.

> and there is no way to disable this when the hardware or driver is buggy

It is not a mpv job to workaround buggy drivers. And for sure not something we should do preemptively. If this is a problem, we can diagnose the root cause and think of the solution that fits certain case.

> or if I prefer mpv's internal tonemapping

What do you mean? This is precisely what it does now. You can set `--target-` to your liking tonemapping or not and it would signal proper image characteristics based on target_params.

> The display will also be stuck in HDR mode if mpv is killed.

Nothing we can do about it if we get SIGKILL and how this is related to this discussion?

Dudemanguy replied:
It's worth noting that AFAIK there's no way for us to know what the right `target-prim` is on this level unless anybody has any ideas.

Dudemanguy replied:
I think there is some misunderstanding here. This doesn't unconditionally signal anything. By default it will always be SDR because that's what the target_params will be. You have to manually specify `--target-prim` and `--target-trc` for this to even do anything. It would be nice if `--target-colorspace-hint` worked, but I have no clue how you would be able to find the right prim for the underlying display. As some people testing in this very PR noted, some things needed bt.2020 and others needed display-p3.

A generic mechanism like kasper was talking about earlier to signal colorspaces/metadata might be nice but I don't think it would be useful for drm. At least not in its current state anyway. That would have more practical use for wayland.

Dudemanguy replied:
>My point is that just specifying --target-prim and --target-trc by themselves shouldn't signal the colorspace, because that's also the case for gpu-next when --target-colorspace-hint=no.

But they do?

kasper93 replied:
> My point is that just specifying `--target-prim` and `--target-trc` by themselves shouldn't signal the colorspace, because that's also the case for gpu-next when `--target-colorspace-hint=no`.

That's valid point, but truth is current system in gpu-next is bit broken, because it works only if the colorspace is signaled through libplacebo which works only with Vulkan and DXGI api.

This whole system needs to be refactored, now that we have ability to signal colorspace in meaningful way in mpv. For example currently there is no way to pass real metadata to DRM, sure it will set PQ in target_params, but max_luma/min_luma will be incorrect.

`--target-colorspace-hint=no` with `--target-trc=pq` doesn't make any sense, except maybe for debugging, but this should never be needed. Depending on API used, there are multiple places where the colorspace needs to be set correctly. You cannot set both of those options and expect it to display correctly. Even if your screen is in "HDR" mode, there are thing in-between that might just change that if colorspace is not "set".

It is as if we had `--target-resolution` and `--target-resolution-hint=no` and try to copy 1080p buffer onto 720p surface... 

There has to be implemented negotiation between drm and gpu-next, what is supported and what should be on output. `--target-colorspace-hint=no` can be left for debugging purposes, but I would rename it to `--target-colorspace=<auto|no>`. Where auto would do thing properly and `no` would be SDR fallback for whatever reason. We live in a word where not everything connected to PC is sRGB anymore and this has to be reflected.

Note that vo_gpu is unconditionally signaling HDR with D3D11, but I don't want to go into vo_gpu territory, it is another level of complexity to solve...

> But they do?

They won't signal anything with `--target-colorspace-hint=no`, resulting in broken colors.

Dudemanguy replied:
It's overriden in `apply_target_options`. Calling `pl_swapchain_colorspace_hint` is just what lets libplacebo infer the colorspace or not which is not really relevant here.

kasper93 replied:
> It's overriden in `apply_target_options`. Calling `pl_swapchain_colorspace_hint` is just what lets libplacebo infer the colorspace or not which is not really relevant here.

It is relevant, because without it swapchain will not be reconfigured properly for HDR content. It will just be srgb default one.

kasper93 replied:
> Well I meant for DRM egl though. There is nothing extra libplacebo side should be doing here AFAIK. It completely depends on the drm code.

gpu_next sets teraget_params to properties of the swapchain, so it won't set HDR if colorspace_hint is not enabled/working. So it doesn't work correctly at this moment in time.

BergmannAtmet commented:
Just tried this.

It does enable HDR mode, but the rendering is broken.

mpv screenshot

![test_cap](https://github.com/user-attachments/assets/803f33e8-9718-4cba-acf3-cf420edb2a11)

physical

![test_tv](https://github.com/user-attachments/assets/ad659128-7f91-475a-9c8d-bf96470b02d0)



Also worthy of note, the mode resets correctly after exiting mpv with `vo=gpu-next`, but it doesn't with `vo=gpu` (the greenish mask stays, it's like broken alpha or something).

I could be missing something obvious here of course. I'm not sure.

Dudemanguy commented:
>Also worthy of note, the mode resets correctly after exiting mpv with vo=gpu-next, but it doesn't with vo=gpu (the greenish mask stays, it's like broken alpha or something).

That should hopefully be fixed now. @quietvoid or @juw: can either of you test if this still works correctly for you? It should but maybe I did something wrong while refactoring things.

BergmannAtmet commented:
> That should hopefully be fixed now.

It is.

quietvoid commented:
Still works fine here.

BergmannAtmet commented:
> Just tried this.
> 
> It does enable HDR mode, but the rendering is broken.
> 

Earlier test was with LTS kernel (6.6.58). Tested again with latest (6.11.5). Same issue. So, it's not a kernel issue.

---

Mesa info for completeness:

```
Device:  AMD Radeon RX 570 Series (radeonsi, polaris10, LLVM 18.1.8, DRM 3.54, 6.6.58-1-lts)
Version: 24.2.5
```

---
I tried my test HDR files with the TV's own OS, and they all rendered correctly in HDR mode. So it's supposedly not an output device issue either.

juw commented:
> Still works fine here.

For me it also still works fine.

kasper93 commented:
PR looks good, but I think we should fix gpu-next metadata before merging it. Doesn't have to made bug changes, we can hack it for now, but I don't feel comfortable discarding all HDR metadata and advertise some dummy values.

Dudemanguy commented:
Sorry I know you've already explained this a gazillion times but I'm totally lost on whatever needs to be done to gpu-next.

kasper93 commented:
> Sorry I know you've already explained this a gazillion times but I'm totally lost on whatever needs to be done to gpu-next.

No worries, it is my fault to not explain clearly. I also not sure myself what is the best way to resolve this. It's all about how `target_params` are populated by `gpu_next`. 
https://github.com/mpv-player/mpv/blob/466b5530c9ca140ce70cd4d41295a987e32992e7/video/out/vo_gpu_next.c#L1133-L1141 The important part is that `target` here is the swapchain format
https://github.com/mpv-player/mpv/blob/466b5530c9ca140ce70cd4d41295a987e32992e7/video/out/vo_gpu_next.c#L1027-L1028 and HDR metadata is only present if libplacebo successfully set those HDR metadata on said swapchain.
https://github.com/haasn/libplacebo/blob/118d8106640796d3f2ceb55f8634a32a58a47aa2/src/vulkan/swapchain.c#L314-L315
This value is used to initialize swapchain frame properites and later used by mpv as `target_params`, because this is what is real target.

Now in case of DRM / EGL there is no `colorspace_hint` implemented and libplacebo/gpu_next will always default to sRGB, stripping all HDR metadata in the process. Even if you set `--target-trc` the metadata is already lost and instead the default one of 10000 nits is used after `apply_target_options`.

In short gpu_next has to know that the swapchain can be configured to use 10bit surface and that we suppose to render HDR to it. While `colorspace_hint` will be done "externally". 

You can test that easily with any HDR video that whatever you do, target_params will not contain correct metadata. Which is nominal case should be video metadata passthrough to display unless user sets `target-peak` to override peak values to better match display capabilities.

As to how to fix that, I'm not sure, some things like creating swapchain is on libplacebo side, but if we can make it create high bitdepth swapchain it is good enough for us. But I didn't look at the details, how to connect things together.

kasper93 commented:
> > and HDR metadata is only present if libplacebo successfully set those HDR metadata on said swapchain.
> 
> Is there a way for mpv to know whether this is set successfully or not? IMO what gpu-next should do here is to first try to signal colorspace hint first through libplacebo, and if libplacebo can't do it, use the GPU context/RA's implementation, which will be a new member function of RA API to signal colorspace hint.

Yes, that would be the way. This is mostly what I meant, that we need some negotiation protocol between gpu-next and rest of the word, if it can render "HDR" and if RA should signal it. Frankly it would be way easier if there wasn't API boundary here. Also `colorspace_hint` itself is not that critical, because we can do it by various means, the tricky part is the swapchain configuration that is currently highly dependent on `colorspace_hint` in libplcebo internally.

No, `pl_swapchain_colorspace_hint` doesn't return any feedback. We can infer that HDR is not set from swapchain frame though. But at this point we should just update libplacebo for proper support. And this is what I worry. I wanted to find some way to make this work without api changes, but it is tricky.

BergmannAtmet commented:
> It does enable HDR mode, but the rendering is broken.

Testing this further, this PR causes a general regression, and it's not related to passing HDR data. The same broken rendering happens when passing --target-peak=204, even with SDR files.

So rendering is broken when tone-mapping is not triggered?

Dudemanguy commented:
@BergmannAtmet: Unfortunately I have zero idea what you're experiencing.

BergmannAtmet commented on video/out/drm_common.c:
```diff
@@ -1240,6 +1353,52 @@ double vo_drm_get_display_fps(struct vo_drm_state *drm)
+
+    // For any HDR, the BT2020 drm colorspace is the only one that works in practice.
+    struct drm_atomic_context *atomic_ctx = drm->atomic_context;
+    int colorspace = !force_sdr && pl_color_space_is_hdr(&drm->target_params.color) ?
```
@Dudemanguy

I took a closer look.

`pl_color_space_is_hdr()` is defined in libplacebo as:

``` c
bool pl_color_space_is_hdr(const struct pl_color_space *csp)
{
    return csp->hdr.max_luma > PL_COLOR_SDR_WHITE ||
           pl_color_transfer_is_hdr(csp->transfer);
}
```

`PL_COLOR_SDR_WHITE` is 203. So that function will return true if you simply set `target-peak` to any value > 203, even with SDR files. Playing a tone-mapped HDR file, which usually involves setting target-peak to values like 400, will also make that function return true.

That was the source of confusion for me. Maybe `pl_color_transfer_is_hdr()` should be used instead here.

I get broken rendering on my device when  "Colorspace" is set to any of

```
    DRM_MODE_COLORIMETRY_OPRGB = 7,
    DRM_MODE_COLORIMETRY_BT2020_RGB = 9,
    DRM_MODE_COLORIMETRY_BT2020_YCC = 10,
```

This is true even without switching to HDR mode. EDID info mentions support for the later two only.

No issue when "Colorspace" is explicitly set to `DRM_MODE_COLORIMETRY_DEFAULT` or `DRM_MODE_COLORIMETRY_BT709_YCC`. Other values don't apply.

BergmannAtmet replied:
> But if your display advertises support for BT2020_RGB and it doesn't actually work, not sure what we can do?

Yeah, I'm aware the problem lies elsewhere, not with mpv. Just reporting my testing results.

In fact, I'm happy to report that I tested with my laptop and the same TV (same input and same HDMI wire too), and got HDR mode enabled with no broken rendering.

For reference, laptop's GPU is a Turing GTX 1650, and I use the nvidia driver (560.35.03).

So now we know that the issue is probably caused by a bug in the open AMD driver stack (POLARIS10 generation at least).

BergmannAtmet commented:
The rendering issue I had is confirmed to be an [AMD HDMI bug](https://gitlab.freedesktop.org/drm/amd/-/issues/3358). So this PR is good to go IMHO.

Dudemanguy commented:
I think the drm code is good. The only problem is what kasper mentioned earlier about gpu_next not passing the correct metadata to the backend. I haven't taken the time to look into that to see if there's a nice way to do that without libplacebo api changes.

Dudemanguy commented:
I added a couple of commits that essentially passes the colorspace from the source directly to the target params and allows the backend to handle it (i.e. `drm` in this case). I figured just skipping the entire colorspace hint API made the most sense because it wasn't really designed with something like drm in mind. Not sure if more is technically needed but the correct metadata should be passed now.

I also added a commit to enable `--target-colorspace-hint` by default because it makes no sense to disable it.

kasper93 commented on video/out/opengl/context_drm_egl.c:
```diff
@@ -708,6 +708,11 @@ static bool drm_egl_reconfig(struct ra_ctx *ctx)
 
+static bool drm_egl_pass_colorspace(struct ra_ctx *ctx)
+{
+    return true;
```
It is fine for now, but I would expect it to be smarter and actually return true only if we are actually able to signal colorspace and is supported by display. Also this could instead of bool return supported modes.

Dudemanguy replied:
I thought about doing that but drm already handles that on its own anyway so it seemed unnecessarily smart.

kasper93 replied:
If you would return `false` from here, gpu-next would nicely fallback to sRGB. By returning always `true` it will always try to output HDR PQ, even if not wanted.

Dudemanguy replied:
Added a little something which I think should work.

kasper93 commented on video/out/vo_gpu_next.c:
```diff
@@ -1001,8 +1005,8 @@ static void draw_frame(struct vo *vo, struct vo_frame *frame)
 
-    struct pl_swapchain_frame swframe;
     struct ra_swapchain *sw = p->ra_ctx->swapchain;
+    struct pl_swapchain_frame swframe;
```
This change is not really needed.

Dudemanguy replied:
Oops some weird leftover that wasn't supposed be here

kasper93 commented on video/out/vo_gpu_next.c:
```diff
@@ -1001,8 +1005,8 @@ static void draw_frame(struct vo *vo, struct vo_frame *frame)
         pl_swapchain_colorspace_hint(p->sw, NULL);
```
Make sure libplacebo doesn't duplicate our job. guard `pl_swapchain_colorspace_hint` with `if (!pass_colorspace)` below. It should be enough for now. We might need to ensure that swapchain format is not set to 8-bits, when we try to pass colorspace, this what libplacebo would do internally.

kasper93 reviewed (approved):
Looks good. At least with gpu-next it will do the job.

kasper93 commented on video/out/vo_gpu_next.c:
```diff
@@ -990,16 +990,21 @@ static void draw_frame(struct vo *vo, struct vo_frame *frame)
     if (p->next_opts->target_hint && frame->current) {
-        struct pl_color_space hint = frame->current->params.color;
+        hint = frame->current->params.color;
+        if (p->ra_ctx->fns->pass_colorspace && p->ra_ctx->fns->pass_colorspace(p->ra_ctx))
```
> I think this should be changed so that pass is only attempted if this function is called, when target-colorspace-hint is set, and don't attempt otherwise. This makes the behavior the same as pl_swapchain_colorspace_hint.

I think it doesn't make sense to make it conditional if it is already based on `target_colorspace` structure. Also it makes little sense to not signal colorspace. `target-colorspace-hint` should be removed, but that's outside of the scope of this PR. It is job for the player to render and send image in proper colorspace and metadata. IMHO sRGB fallback is ancient artifact, many monitors doesn't even have sRGB mode, so it is wrong.

Dudemanguy replied:
The idea behind the naming was that vo_gpu_next passes the colorspace directly to the backend without going through libplacebo. Hence `pass_colorspace`. If you have a better name, I don't mind changing it.

Also it should already only be attempted if `--target-colorspace-hint` is set? Or am I misunderstanding what you mean here?

Dudemanguy replied:
>vo_drm_set_hdr_metadata is always called

Sure but it should effectively be a no-op if you aren't using `--target-colorspace-hint`.

Dudemanguy replied:
>This can change display behavior.

Does it? I would expect drivers to no-op the hdr metadata parameters if you aren't trying to use the HDR mode. It would be pretty broken if they actually tried to do anything with it. But OK it doesn't hurt to be paranoid I guess.

>vo_gpu

This has many problems in this area that I'm not going to bother fixing.

Dudemanguy replied:
Hmm but this proposed change is actually pretty annoying in reality to implement because it messes up the "restore back to sdr" logic.

Dudemanguy replied:
@na-na-hi: Yeah the additional code complexity is not worth it IMO unless you are aware of an actual concrete problem. Setting the metadata blob with SDR works exactly as expected for me. Setting the blob whenever the params change from the VO is the cleanest and best way to handle it. If you want to make vo_gpu aware of various `--target` options, that's fine but I'm not going to do it in this patch.

Dudemanguy commented on video/out/drm_common.c:
```diff
@@ -588,6 +617,83 @@ static bool setup_crtc(struct vo_drm_state *drm, const drmModeRes *res)
+    while (*exts && !(cta = di_edid_ext_get_cta(*exts++)));
+
+    if (!cta) {
+        MP_WARN(drm, "Failed to get CTA-861 extension block.\n");
```
Probably just a mistake on my part. Should be verbose like the rest.

kasper93 replied:
Maybe also make it "not available" instead of "failed"? The later implies some error, while I think it is expected outcome.

Dudemanguy replied:
Made it say `Unable to find CTA-861 extension block.` now

Dudemanguy commented:
Dropped one additional message from warn down to verbose.

hooke007 commented:
https://github.com/mpv-player/mpv/commit/23843b4aa594dc8c885575f3d237cde3c29398a2#comments

hooke007 commented:
Mac doesn't prepare well for this option neither.
https://github.com/mpv-player/mpv/issues/12666

kasper93 commented:
> Also in real world, the experience of countless devices's hdr passthrough is just a disaster. Retonemapping to SDR is always the better choice in recent years.

This is 100% true, but we cannot be stuck behind assumptions like that. I think there is no correct way, there are multiple of factors that may make you use one or the other mode. If we want to stay with sRGB forever why even bother implementing all this protocols. Either way, it is a choice, many people are requesting passthrough, because they bought new fancy HDR displays. You can opt-out if needed.
